scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketTransform.java

```diff
@@ -0,0 +1,273 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.extensions.smb.BucketShardId.BucketShardIdCoder;
+import org.apache.beam.sdk.extensions.smb.FileOperations.Writer;
+import org.apache.beam.sdk.extensions.smb.SMBFilenamePolicy.FileAssignment;
+import org.apache.beam.sdk.extensions.smb.SortedBucketSink.WriteResult;
+import org.apache.beam.sdk.extensions.smb.SortedBucketSource.BucketedInput;
+import org.apache.beam.sdk.extensions.smb.SortedBucketSource.SourceSpec;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.io.fs.ResourceIdCoder;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Reshuffle;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.transforms.join.CoGbkResultSchema;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions;
+
+public class SortedBucketTransform<FinalKeyT, FinalValueT> extends PTransform<PBegin, WriteResult> {
+  private final SMBFilenamePolicy filenamePolicy;
+  private final ResourceId tempDirectory;
+  private final FileOperations<FinalValueT> fileOperations;
+  private final Class<FinalKeyT> finalKeyClass;
+  private final List<BucketedInput<?, ?>> sources;
+  private final BucketMetadata<FinalKeyT, FinalValueT> bucketMetadata;
+  private final TransformFn<FinalKeyT, FinalValueT> transformFn;
+
+  public SortedBucketTransform(
+      Class<FinalKeyT> finalKeyClass,
+      BucketMetadata<FinalKeyT, FinalValueT> bucketMetadata,
+      ResourceId outputDirectory,
+      ResourceId tempDirectory,
+      String filenameSuffix,
+      FileOperations<FinalValueT> fileOperations,
+      List<BucketedInput<?, ?>> sources,
+      TransformFn<FinalKeyT, FinalValueT> transformFn) {
+    this.filenamePolicy = new SMBFilenamePolicy(outputDirectory, filenameSuffix);
+    this.tempDirectory = tempDirectory;
+    this.fileOperations = fileOperations;
+    this.finalKeyClass = finalKeyClass;
+    this.sources = sources;
+    this.transformFn = transformFn;
+    this.bucketMetadata = bucketMetadata;
+  }
+
+  @Override
+  public final WriteResult expand(PBegin begin) {
+    final SourceSpec<FinalKeyT> sourceSpec = SortedBucketSource.getSourceSpec(finalKeyClass, sources);
+
+    Preconditions.checkArgument(
```

Comment:
on second thought, I guess the user could specify a larger number of buckets--since at merge time we have access to the key bytes, we could just call `BucketMetadata#getBucketId(byte[] keyBytes)` using the `BucketMetadata` that's supplied for the output, then open a Writer for that bucket file... we would potentially have a lot of open file handles per DoFn though. but it could be a nicer option than offering the round robin shardingg thing.

Reply:
Either way we'll have to keep all bucket/shard files open right? Could be an optimization for later?

Reply:
Yeah, I was kind of thinking we could offer this instead of the sharding? At least for a first pass? wdyt?

Reply:
Yeah on second thought this should be better than sharding because it enables more parallelism downstream given that the data size is likely growing? :+1:

Reply:
yep! cool, I'll update the PR soon with the change.

Reply:
done! verified in unit test/dataflow. I think this PR is ready to merge